cmd/agent/daemon/state/events_pipeline.go

```diff
@@ -19,125 +20,278 @@ func (c *Controller) runEventsPipeline(ctx context.Context) error {
 	c.log.Info("running events pipeline")
 	defer c.log.Info("events pipeline done")
 
+	var currentEventsCount int
+	lastFlushedAt := c.nowFunc()
+	groups := map[uint64]*castpb.ContainerEvents{}
+
+	ticker := time.NewTicker(c.cfg.EventsFlushInterval)
+	defer ticker.Stop()
+
+	send := func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+
+		batch := &castpb.ContainerEventsBatch{}
+		for _, group := range groups {
+			if len(group.Items) > 0 {
+				batch.Items = append(batch.Items, group)
```

Comment:
Nitpick: we either should keep the `batch` event around and reset the items after each send, or we should precalc the size of the `batch.Items` slice and preallocate it. This will cause a lot of memory churn.